lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -303,32 +330,85 @@ def _compute_settings(
         1) Remove the entries to be deleted
         2) Add entries, if available
+        Returns True if a configuration change was performed on opensearch.yml only
+        and a restart is needed.
+
+        Executes the following steps:
+        1) Tries to manage the keystore
+        2) If settings API is available, tries to manage the configuration there
+        3) Inserts / removes the entries from opensearch.yml.
+
+        Given keystore + settings both use APIs to reload data, restart only happens
+        if the configuration files have been changed only.
 
-        Returns True if a configuration change was performed.
+        Raises:
+            OpenSearchKeystoreNotReadyYetError: If the keystore is not yet ready.
         """
-        self._keystore.delete(config.secret_entries_to_del)
-        self._keystore.add(config.secret_entries_to_add)
-        if config.secret_entries_to_del or config.secret_entries_to_add:
-            self._keystore.reload_keystore()
+        keystore_ready = True
+        cluster_settings_changed = False
+        try:
+            # If security is not yet initialized, this code will throw an exception
+            self._keystore.delete(config.secret_entries_to_del)
+            self._keystore.add(config.secret_entries_to_add)
+            if config.secret_entries_to_del or config.secret_entries_to_add:
+                self._keystore.reload_keystore()
+        except (OpenSearchKeystoreNotReadyYetError, OpenSearchHttpError):
+            # We've failed to set the keystore, we need to rerun this method later
+            # Postpone the exception now and set the remaining config.
+            keystore_ready = False
 
         current_settings, new_conf = self._compute_settings(config)
-        if current_settings == new_conf:
-            # Nothing to do here
-            logger.info("apply_config: nothing to do, return")
-            return False
+        if current_settings and new_conf and current_settings != new_conf:
+            if config.config_entries_to_del:
+                # Clean to-be-deleted entries
+                self._opensearch.request(
+                    "PUT",
+                    "/_cluster/settings?flat_settings=true",
+                    payload={"persistent": {key: "null" for key in config.config_entries_to_del}},
+                )
+                cluster_settings_changed = True
+            if config.config_entries_to_add:
+                # Configuration changed detected, apply it
+                self._opensearch.request(
+                    "PUT",
+                    "/_cluster/settings?flat_settings=true",
+                    payload={"persistent": config.config_entries_to_add},
+                )
+                cluster_settings_changed = True
 
-        # Update the configuration
+        # Update the configuration files
         if config.config_entries_to_del:
             self._opensearch_config.delete_plugin(config.config_entries_to_del)
```

Comment:
why is there a mixing between API calls (above) and the same on the opensearch.yml level? how do you differentiate between the ones related to the file and the ones of the API?

Reply:
Every plugin change from the plugin manager is persisted on file.

Reply:
In short, I am not right now making any differences. What I am doing there is making sure the configs are persisted in file and, if possible, we apply everything via API.